### DISCUSSION TITLE
How do I set response headers in a CatchBoundary?

### DISCUSSION THREAD

**Discussion by jamesarosen** [remix-run/remix#5058] [Q&A]
Based roughly on https://remix.run/docs/en/v1/route/loader#throwing-responses-in-loaders, 

I have some centralized authentication code:
```js
async function ensureSession(request) {
  const session = await getSession(request.headers.get('cookie'))
  const user = session?.get('user')
  if (user == null) {
    const headers = new Headers()
    headers.append('set-cookie', returnToCookie.serialize(request.url))
    throw new Response("Unauthorized", { headers, status: 401 })
  }

  return user
}
```
and a CatchBoundary that captures that `Response` via `useCatch`. When I try to go to a page that requires a session, I get a 401, as expected.

I'd like that 401 response to include a `Set-Cookie` header with the return-to-after-sign-in URL, but I can't figure out how to do that. I've tried this:

```tsx
export const loader: LoaderFunction = async (request) => {
  return ensureSession(request)
};

export function CatchBoundary() {
  const caught = useCatch()
  if (caught.status === 401) return caught

  return <p>Whoops!</p>
}
```

That doesn't work because `CatchBoundary` needs to return a `ReactNode`, not a `Response`.

I could `.catch(…)` in my `loader` and return a `Response` there, but that defeats the point of having the `CatchBoundary` to centralize error-handling.

The best idea I have is to write a higher-order loader function that wraps any route's `loader` in 401-aware error handling.

Am I missing something? Is there some _completely different way of accomplishing this?

**Comment by kiliman** [comment]
First of all, `<CatchBoundary/>` is a UI component. It just renders based on the caught response. Logic should be handled in loaders and actions.

I would recommend that `ensureSession()` throws a redirect to your login page with the _returnUrl_ as a search param.

```ts
async function ensureSession(request) {
  const session = await getSession(request.headers.get('cookie'))
  const user = session?.get('user')
  if (user == null) {
    const url = new URL(request.url)
    throw redirect(`/login?returnUrl=${url.pathname}${url.search}`)
  }

  return user
}
```
Then in your login action, check for `returnUrl` and redirect there if present or to your home page.

```ts
// routes/login.tsx
export async function action({request}: ActionArgs) {
  const formData = await request.formData()
  // verify login...
  
  const url = new URL(request.url)
  let redirectUrl = url.searchParams.get('redirectUrl')
  // only redirect to our own routes, not external
  if (!redirectUrl?.startsWith('/')) {
    redirectUrl = '/'
  }
  return redirect(redirectUrl)
}
```

**Reply by jamesarosen** [reply]
> I would recommend that `ensureSession()` throws a redirect to your login page with the `returnUrl` as a search param.

That works great for initial page loads:

1. with no session, user goes to `myapp.com/private`
1. `ensureSession` redirects to `myapp.com/login?return-to=%2Fprivate`
1. user logs in
1. user is redirected to `/private` 🥳 

That doesn't work at all for AJAX requests:

1. with no session, user goes to `myapp.com/`
1. user clicks an action like "save to my bookmarks"
1. `<Bookmark>` component makes a `POST /bookmarks`
1. `ensureSession` redirects to `myapp.com/login?return-to=%2Fprivate`
1. browser automatically fetches `myapp.com/login?return-to=%2Fprivate`
1. `<Bookmark>` component receives login HTML instead of JSON 💥 

It _could_ work well for page transitions:

1. with no session, user goes to `myapp.com/`
1. user clicks link to `/private`
1. `ensureSession` redirects to `myapp.com/login?return-to=%2Fprivate`
1. user logs in
1. user is redirected to `/private` 🥳 

But that assumes that the desired behavior is going to a login page. In our app, it makes more sense to display a sign-up/login modal over the current content. A page transition would be disruptive to the flow.

So at minimum, I need two versions of `ensureSession`: one that redirects for use in HTML routes and one that returns a 401 for API route actions.

**Reply by kiliman** [reply]
It appears you're talking about resource routes. Yes, I typically have an `api` folder for resource routes, and can test to see if I should return a redirect or a JSON response.

I don't typically throw a response from resource routes because the caller can't handle the response. Since resource routes are not UI, it seems odd to have errors end up in `CatchBoundary`.

So if your `<Bookmark>` component is calling a resource route, you should either return success or an error via `return json({ success | error}, { status: 200 | 401 | etc })`

Then check `fetcher.data` to see what the result was and act accordingly. For example, I'm not sure what the expectation is if somebody tries to save to bookmarks if there's no session.

**Reply by jamesarosen** [reply]
> So if your `<Bookmark>` component is calling a resource route, you should either return success or an error via return `json({ success | error}, { status: 200 | 401 | etc })`

It absolutely should! I'm just trying to centralize authentication (and authorization) as much as possible. It sounds like the most centralized I can get is

1. one `ensureSession` for routes that service page loads (via location change, including form submission), which throws a redirect, and
2. one `ensureSession` for resource routes (via fetch), which returns or throws a 401

**Reply by jamesarosen** [reply]
> I'm not sure what the expectation is if somebody tries to save to bookmarks if there's no session.

It would look something like

```jsx
const saveBookmark = useCallback(() => {
  fetch(`/api/bookmarks', { method: 'post', body: JSON.stringify({ post_id: post.id }))
  .then(res => {
    if (res.status < 400) { setIsBookmarked(true) }
    else if (res.status === 401) { setSignInModalVisible(true) }
    else { setError(res) }
  })
}, [post])
```

Of course, if there are other `fetch` calls that require a session, I would put the sign-in-modal-visibility flag in a context and wrap `fetch` (or `useFetcher`) in something that knows about that context.

**Reply by kiliman** [reply]
Ah. I guess the confusion is that you're working outside the Remix framework. If you're going to use `fetch` then you can do whatever you want. If you're using Remix *fetchers*, then it's best to stick with established patterns.

**Reply by jamesarosen** [reply]
The product requirement is (assuming the browser has JS) that a modal pops up with no page transition. Given that requirement, I think we _have_ to work outside the Remix framework. The Fetcher returned from `useFetch` doesn't let us capture the 401 response within the scope of the component and `CatchBoundary` replaces the current route component.

I'd love to use `useFetcher` for this, but I don't think it's possible. I can imagine two different ways of improving Remix to support this use-case:

 1. A way of making `CatchBoundary` _additive_ to its route component. This would, for example, let apps show a toast message over the current context.
 2. A way of processing a `useFetcher` response within the component that called `fetcher.submit()`, preventing the error from bubbling up to the `CatchBoundary` altogether

**Reply by sergiodxa** [reply]
You could do `return json({ error: "something" }, { status: 401 })` and then use it with useFetcher

```ts
let { data } = useFetcher()
let failed = data?.error === "something"
```

If you don't throw the response the CatchBoundary will not show up and you can still set a 4xx or 5xx status code

**Reply by sergiodxa** [reply]
I have built a route that does something like you mention, if there's JS show the result in a modal, if there's no JS show it inline.

I used Form instead of useFetcher and I post to the same route, then I use the `useHydrated` hook from Remix Utils, if the app is hydrated I show a dialog if not I show the result inline on the page.

**Reply by jamesarosen** [reply]
> You could do `return json`… If you don't throw the response the CatchBoundary will not show

Aha! I think that distinction is what I was looking for.

I really like [Rails's `rescue_from` pattern](https://guides.rubyonrails.org/action_controller_overview.html#rescue-from). It lets you `raise` (`throw`) from anywhere, then `rescue` (`catch`) locally if you want or bubble up to the controller (route) for default handling. Adapting that to resource routes in Remix, I might do something like

```js
// app/util/session.js
export interface ActionFunctionWithUser {
    (args: ActionFunctionArgs & { user: User }): Promise<Response> | Response | Promise<any> | any;
}

/**
 * Wraps a Resource Route action function with session-awareness.
 */
export function withSessionEnsured(action: ActionFunctionWithUser): ActionFunction {
  return function loaderWithSessionEnsured({ request, ...rest }) {
    const session = await getSession(request.headers.get('cookie'))
    const user: User | null = session?.get('user')

    if (user == null) {
      const headers = new Headers()
      headers.append('set-cookie', returnToCookie.serialize(request.headers.get('referer')))
      return json({ error: "No session" }, { headers, status: 401 })
    }

    return loader({ request, ...rest, user })
  }
}

// app/routes/bookmark.ts
export const loader = withSessionEnsured(async ({ request, user }) => {
  // do something with request and user
});
```
